Implement toolbar extension mechanism for providers

martinpovolny opened the pull request:
### Context

We want to allow the provider authors to extend ManageIQ UI with Ansible playbooks and roles. ManageIQ core provides the capability to execute an Asible playbook from the provider repository. The ManageIQ UI provides a way to call that playbook from a toolbar. Before calling the playbook additional parameters can be collected using a custom dialog.

Related UI PRs:
 * https://github.com/ManageIQ/manageiq-ui-classic/pull/4267
 * https://github.com/ManageIQ/manageiq-ui-classic/pull/4231

### Goal of this PR

This allows providers to extend existing toolbars by creating classes such as:

`app/helpers/manageiq/providers/amazon/toolbar_overrides/ems_cloud_center.rb` in the `manageiq-providers-amazon` repo.

```ruby
module ManageIQ
  module Providers
    module Amazon
      module ToolbarOverrides
        class EmsCloudCenter < ::ApplicationHelper::Toolbar::Override
          button_group('magic', [
            button(
              :magic,
              'fa fa-magic fa-lg',
              t = N_('Magic'),
              t,
              :data  => {'function'      => 'sendDataWithRx',
                         'function-data' => {:controller     => 'provider_dialogs', # this one is required
                                             :button         => :magic,
                                             :modal_title    => N_('Create a Security Group'),
                                             :component_name => 'CreateAmazonSecurityGroupForm',
                                              }.to_json},
              :klass => ApplicationHelper::Button::ButtonWithoutRbacCheck),
            button(
              :magic_dialog,
              'fa fa-magic fa-lg',
              t = N_('Magic'),
              t,
              :data  => {'function'      => 'sendDataWithRx',
                         'function-data' => {:controller => 'provider_dialogs', # this one is required
                                             :button     => :magic_dialog}.to_json},
              :klass => ApplicationHelper::Button::ButtonWithoutRbacCheck),
          ])
        end
      end
    end
  end
end

```

Any toolbar can be overridden but we focus on providers on the entity detail screen in this PR. Examples: 
 * Instance detail screen, 
 * Cloud provider detail screen (the example above).

For these (detail) screens. The toolbar is not only extended, but the proper extension is selected by matching `@record` with the toolbar's class.

For screens where there is no `@record` all available extensions are loaded.

### How it is done?

Toolbar base class is reorganized as follows:

 * `Toolbar::Basic` is moved to a base class `Toolbar::Base`
 * `Toolbar::Override` (practically `Toolbar::Base`) is where the provider
	extentions should inherit from.
 * `Toolbar::Basic` newly includes the extention mechanism.

`ToolbarBuilder` now passed the `@record` to the toolbar `definition` method (now a method). And it filters the available extensions by matching the record's class with the extension class (the provider base module).


Example provider PR: https://github.com/ManageIQ/manageiq-ui-classic/pull/4231

martinpovolny commented:
to get rid of the `require` statement, I can add
```
config.autoload_paths << Rails.root.join("app", "toolbar_overrides").to_s
```
to `application.rb` in manageiq (core).

With that:
```
[9] pry(#<ApplicationHelper::Toolbar::EmsCloudCenter>)> ManageIQ::Providers::Amazon::ToolbarOverrides::EmsCloudCenter                                                                                              
NameError: uninitialized constant ManageIQ::Providers::Amazon::ToolbarOverrides                          
from (pry):1:in `block in extension_classes'
```

I believe the autoloades splits the class name on "::" and resolves each module in a turn and it does not see a module name `ManageIQ::Providers::Amazon::ToolbarOverrides` until the file with `ManageIQ::Providers::Amazon::ToolbarOverrides::EmsCloudCenter` is loaded:

`manageiq-providers-amazon/app/toolbar_overrides/ems_cloud_center.rb` :
```
module ManageIQ                                                                                     
  module Providers                                                                                  
    module Amazon                                                                                   
      module ToolbarOverrides                                                                       
        class EmsCloudCenter < ::ApplicationHelper::Toolbar::Override                               
          button_group('magic', [  
...
```

I can "fix" that:
```
[10] pry(#<ApplicationHelper::Toolbar::EmsCloudCenter>)> module ManageIQ::Providers::Amazon::ToolbarOverrides; end                                                                                                 
=> nil                                              
[11] pry(#<ApplicationHelper::Toolbar::EmsCloudCenter>)> ManageIQ::Providers::Amazon::ToolbarOverrides::EmsCloudCenter 
```

So the problem is the `:ToolbarOverrides::` in the class name. But I'd like to keep i there.

What's the best way to fix that?

martinpovolny commented:
ping @Fryguy

Fryguy commented on app/helpers/application_helper/toolbar/basic.rb:
```diff
@@ -44,9 +42,43 @@ def separator
+        instance = [plugin.name.sub('::Engine',''), 'ToolbarOverrides', self.class.name.to_s.split('::').last].join('::')
+        instance.constantize
+      end
+    end.compact
```
I think this can all live in the Vmdb::Plugins class itself, but this is mostly pretty good :+1:

Fryguy commented on app/helpers/application_helper/toolbar/override.rb:
```diff
@@ -0,0 +1,5 @@
+class ApplicationHelper::Toolbar::Override < ApplicationHelper::Toolbar::Basic
+  def definition
+    @definition
+  end
```
`attr_reader :definition` seems better.

Fryguy commented on app/helpers/application_helper/toolbar/basic.rb:
```diff
@@ -44,9 +42,43 @@ def separator
+      @loaded = true
+    end
+
+    @definition
```
Prefer something like `@definition ||= load_overrides` and then have load_overrides just return the thing.

Fryguy replied:
Ugh...I see that `custom_content` injects itself via the ivars...might be better if those methods all go through the definition accessor.  That way they are loaded "first".  I guess I don't understand the DSL in this class and what is trying to be solved, so perhaps my suggestions should be ignored 😁

martinpovolny replied:
I will be redoing that all. So while your comments make sense, I'll just dump the code and do it differently once the autoloading works.

Fryguy commented on app/helpers/application_helper/toolbar/basic.rb:
```diff
@@ -44,9 +42,43 @@ def separator
+      name = "#{name}.rb"
+      if File.exist?(name)
+        require name
+        instance = [plugin.name.sub('::Engine',''), 'ToolbarOverrides', self.class.name.to_s.split('::').last].join('::')
```
> `plugin.name.sub('::Engine','')`

Slightly better: `plugin.name.chomp("::Engine")`

> self.class.name

What is an example of `self.class.name` here?  Trying to understand what this does....I think it might be easier with a simple .gsub of the middle section, but I'm not sure.

Fryguy commented on app/helpers/application_helper/toolbar/basic.rb:
```diff
@@ -44,9 +42,43 @@ def separator
+    end.compact
+  end
+
+  def load_overrides
```
probably should be a private method.

Fryguy commented on app/helpers/application_helper/toolbar/basic.rb:
```diff
@@ -44,9 +42,43 @@ def separator
+
   def initialize
+    @loaded = false
     @definition = {}
```
Unset both of these and let the accessor deal with the lazy loading on first access.

slemrmartin commented:
@martinpovolny I'm not sure what exactly you need.
It seems like you want to make dependency from core to amazon and use it from ui gem?
With the autoload_paths you mentioned above in core's application.rb, you can use `::ToolbarOverrides::..`, without provider prefix

martinpovolny commented:
@slemrmartin: What I need to do:

We have toolbar classes in the form of `ApplicationHelper::Toolbar::CloudTenantCenter`. These toolbar classes contain a tooolbar definition. `CloudTenantCenter` is what I call here a *toolbar_name*.

The providers will have a way of extending the toolbar class. So I need to query and load class in the form of: 

`<provider_namespace>::ToolbarOverrides::<toolbar_name>`

In case of Amazon provider and the CloudTenantCenter toolbar it would be:

`ManageIQ::Providers::Amazon::ToolbarOverrides::CloudTenantCenter`

I have no problem doing that with a require statement:

```
    Vmdb::Plugins.collect do |plugin|                                                               
      name = plugin.root.join('app/toolbar_overrides', self.class.name.to_s.split('::').last.underscore)
      if File.exist?("#{name}.rb")
        require "#{name}.rb"                                                                  
        instance = [plugin.name.chomp('::Engine'), 'ToolbarOverrides', self.class.name.to_s.split('::').last].join('::')
        instance.constantize                                                                        
      end                                                                                           
    end.compact   
```

And placing the toolbar override classes in the provider repo under 'app/toolbar_overrides'. That works just fine and I am happy.

However I'd like to do that w/o that require statement with the Rails autoloading mechanism.

I am looking for a way
 1. how to place the files with the classes
 1. how to query the existing classes (so that I can merge the toolbar definitions)
 1. how to name the classes and wrapping modules etc.

To get the same or similar effect but w/o the `require` statement using the Rails autoloader.

So far I have no success. I am obviously structuring the modules and classes and files in a way that is not  compatible with how the autoloading works. Help welcome.

slemrmartin commented:
@martinpovolny I think you need to follow namespaces when building directory structure - for autoloader
I think this should work:
- 
in plugin manageiq-providers-amazon:
/lib/manageiq/providers/amazon/engine.rb
```
...
class Engine < ::Rails::Engine
...
  config.autoload_paths << File.expand_path('../../../../app/toolbar_overrides', __FILE__)
```
then you can have:
/app/toolbar_overrides/manageiq/providers/amazon/toolbar_overrides/cloud_tenant_center.rb
```
class ManageIQ::Providers::Amazon::ToolbarOverrides::CloudTenantCenter
end
```

martinpovolny commented:
Problem solved. Thx @slemrmartin, @Fryguy, @Ladas !

martinpovolny commented on app/helpers/application_helper/toolbar/basic.rb:
```diff
@@ -1,66 +1,34 @@
+    # "ManageIQ::Providers::Amazon::ToolbarOverrides::EmsCloudCenter" is a match for
+    #   "ManageIQ::Providers::Amazon::CloudManager
+    extension_classes.find_all do |ext|
+      ext.name.split('::')[0..-3] == record.class.name.split('::')[0..-2]
```
@Ladas, @slemrmartin : Do we have a better way to do the matching?

Ladas replied:
You compare just the ```ManageIQ::Providers::Amazon```, right? So maybe 
```ext.name.split('::')[0..2] == record.class.name.split('::')[0..2]``` looks less fragile

Not sure about a better way

martinpovolny commented:
@himdel, can you, please, review and merge this one?

Fryguy commented:
I would have liked a second chance to review this, considering I had issues 😕

martinpovolny commented:
@Fryguy : sorry for that, will address any feedback in a separate PR.

mzazrivec commented:
@Fryguy Whups, sorry about that. I'm sure we can address any problems in future PRs.
